### DISCUSSION TITLE
Polling 5000 tags

### DISCUSSION THREAD

**Discussion by Dhruva0707** [ottowayi/pycomm3#131] [Q&A]
Reading tags fast,
Hello i need to poll about 5000 tags every 300ms. 
Is it possible? 

There are some objects which are not completely accesible, so i have to poll subtags in them individually

**Comment by ottowayi** [comment]
That's something you're going to have to try out and see if it will work for you, there are a ton of factors that could influence how fast you can read.  What type of tags are you wanting to scan?  You can fit many more simple atomic types like DINT, INT, REAL into a single request, than say strings or UDTs.  If you're reading SINTs, you can fit ~4x as many in a single request as you can with DINTs.  How old is your hardware?  A new L85 is going to be a lot faster than old L55 or an ENBT is much slower than an EN2T.  Does you PLC support large packets?  Older PLCs cannot handle packets larger than 500 bytes, while new one support up to 4000 bytes.  Large packets means less requests and less time overall.  Another feature of newer PLCs is the use of Instance Ids in the request, instead of encoding the tag name in the packet, it uses a short internal id number. If you PLC doesn't support Instance IDs, the length of your tag names will affect how many request you can include in a packet. 

That said, I was able to do 3000 at ~160ms average on a really old CompactLogix, so it should be doable but there's too many unknown for me to say without trying it out on your system.

**Reply by Dhruva0707** [reply]
How do we read using the instance id of internal tags? 

Note: The plc we are using is a Control Logix-1756-L72

**Reply by Dhruva0707** [reply]
Another question, is it possible to make parallel connections to the plc and request the data from each parallel connection?

**Reply by ottowayi** [reply]
1 - it is automatic, as long as your L72 is above v21 it should work. It is part of the driver initialization, during the `open()` call.  You can check the `plc._cfg['use_instance_ids']` setting after calling the open method. 

2 - Yes, but you need to do that.  Each driver is a single client/connection, but you can use multiple drivers in parallel. 

Tip:  If you do open multiple drivers, you can save startup by only uploading the tag list in one driver and sharing it to the others.

```
plc1 = LogixDriver('1.1.1.1')
plc1.open()

plc2 = LogixDriver('2.2.2.2', init_tags=False)
plc2._tags = plc1.tags
plc2.open() 

```

**Reply by Dhruva0707** [reply]
thanks a lot!

**Reply by lil-sahil** [reply]
Hi, just want to clarify, if I would like to read the value of 1 tag continuously (lets say every second), is the correct way to utilize a while loop like so:

while True:
   with pycomm3.LogixDriver( ip_adress ) as plc:
      print( plc.read( '{}'.format( tag_name ) ) )

Or do you recommend another another way of doing this?

Essentially, what I am trying to do is integrate a Raspberry Pi to take a picture every time the sensor connected to the PLC is turned on.
 
Thanks!

**Reply by ottowayi** [reply]
That method (using the `with` block) will open/close a connection every time, it might be fine on a newer hardware but I've definitely seen issues on old devices like ENBTs, especially if you're not sleeping at all between reads.  What you want to do is open the connection beforehand, like:

```
plc = pycomm3.LogixDriver(ip_address)
plc.open()

while True:
    print(plc.read(tag_name))

plc.close()
```

**Reply by lil-sahil** [reply]
Thanks!

**Reply by Ajility97** [reply]
@ottowayi thanks for the info - I'm trying to read ~59,000 tags in one go using a ControlLogix 5572. Currently, pycomm3 can do it in ~12 seconds out of the box:

```
def poll(self):
	tags = self._getMonitoringTagNamesAsList()
	if len(tags) > 0:
		import time
		print('Start poll')
		start = time.time()
		
		polledTags = self.comm.read(*tags)
		
		end = time.time()
		print('Done poll: ', end - start)
		now = datetime.now().strftime('%Y-%m-%d %H:%M:%S')
# Start poll
# Done poll:  12.637672185897827
```

I'm trying to speed this up. My thought was to use threading, so I took this approach inspired by your earlier comment:
```
 def __init__(self, address):
       #.....
	i = 0
        self.comm = LogixDriver(self.address)
	self.threads = 50
	self.commStack = []
	while i < self.threads:
		comm = LogixDriver(self.address, init_tags=False)
		comm._tags = self.comm.tags
		comm.open()
		self.commStack.append(comm)
		i += 1

	# Close all of the PLC connections on program exit
	atexit.register(self._close)

def _close(self):
	for comm in self.commStack:
		comm.close()
		
def task(self, tags):
	comm = self.commStack.pop()
	tags = comm.read(*tags)
	self.commStack.append(comm)

	return tags

def poll(self):
	from multiprocessing.pool import ThreadPool
	pool = ThreadPool(self.threads)
	polledTags = []

	import time
	print('Start poll')
	start = time.time()

	for chunk in chunkArray(self._getMonitoringTagNamesAsList(), 1000):
		polledTags.append(pool.apply_async(self.task, args=(chunk,)))
	pool.close()
	pool.join()

	polledTags = [r.get() for r in polledTags]
	end = time.time()
	print('Done poll: ', end - start)
	now = datetime.now().strftime('%Y-%m-%d %H:%M:%S')

# Start poll
# Done poll:  10.0589120388031
```

There is an improvement of about 3 seconds. Do you see a way to improve this to make it faster? `_cfg['use_instance_ids']` is true in my case.
